source/common/stats/allocator_impl.cc

```diff
@@ -78,27 +68,55 @@ template <class BaseClass> class StatsSharedImpl : public MetricImpl<BaseClass>
   // RefcountInterface
   void incRefCount() override { ++ref_count_; }
   bool decRefCount() override {
+    // We must, unfortunately, hold the allocator's lock when decrementing the
+    // refcount. Otherwise another thread may simultaneously try to allocate the
+    // same name'd stat after we decrement it, and we'll wind up with a
+    // dtor/update race. To avoid this we must hold the lock until the stat is
+    // removed from the map.
+    //
+    // It might be worth thinking about a race-free way to decrement ref-counts
+    // without a lock, for the case where ref_count > 2, and we don't need to
+    // destruct anything. But it seems preferable at to be conservative here,
+    // as stats will only go out of scope when a scope is destructed (during
+    // xDS) or during admin stats operations.
+    Thread::LockGuard lock(alloc_.mutex_);
     ASSERT(ref_count_ >= 1);
-    return --ref_count_ == 0;
+    if (--ref_count_ == 0) {
```

Comment:
It occurred to me that while this change is needed for correctness, the TLS cache structure in thread_local_store.cc where each TLS cache has a map to CounterSharedPtr etc means that when a scope is destructed, all threads will need to acquire the same allocator lock as they destruct.

I have a working-but-not-quite-tested change that moves the ownership in the CentralCache, with the TLS Cache entries just storing refs or pointers. I'd prefer to do that as a follow-up though.

Reply:
That change is now a WiP in https://github.com/envoyproxy/envoy/pull/9564 but I think it'd be better to review this one as is. The main risk is that scope deletion could trigger a flurry of lock contentions but I don't believe there is any steady-state performance drawback to this PR.

#9564 will be much easier to review after this is in.